### PR TITLE
Added Grouped iterable

### PR DESCRIPTION
**Pull request by svendiedrichsen**
Please review my suggestion of grouping an iterable source by using grouping functions. Original issue #595 .

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#642](https://github.com/yegor256/cactoos/issues/642) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#642](https://github.com/yegor256/cactoos/issues/642) is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), here is [why](https://www.0crat.com/footprint/C63314D6Z/451534093). The budget is 15 minutes, see [§4](http://datum.zerocracy.com/pages/policy.html#4). Please, read [§27](http://datum.zerocracy.com/pages/policy.html#27) and when you decide to accept the changes, inform @yegor256[/z](https://www.0crat.com/u/yegor256) (the architect) right in this ticket. If you decide that this PR should not be accepted ever, also inform the architect.

**Comment by amihaiemil** [comment]
@svendiedrichsen Please, link the original issue in the description (without any keyword like "fixes" or "closes", we don't want Github to automatically close the ticket when this PR is merged). 

And can you give an example of how would this iterator be used? Input and output?

**Comment by svendiedrichsen** [comment]
@amihaiemil I have added the link the original ticket to the description. The usage like it is shown in the provided test:
```
new Grouped<>(
    new RangeOf<>(1, 10, value -> ++value),
    new IterableOf<Func<Integer, Boolean>>(
        value -> value % 2 == 0,
        value -> value % 2 != 0
    )
)
```

File: src/main/java/org/cactoos/iterable/Grouped.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ * @since 1.0
+ */
+public final class Grouped<X> extends IterableEnvelope<Iterable<X>> {
+    /**
+     * Ctor.
+     * @param origin The iterable the grouping is based on.
+     * @param funcs The iterable containing grouping functions.
+     */
+    public Grouped(final Iterable<X> origin,
+        final Iterable<Func<X, Boolean>> funcs) {
```
Comment (amihaiemil) [inline src/main/java/org/cactoos/iterable/Grouped.java]: @svendiedrichsen formatting is strange here, I would do it like this:

```java
public Grouped(
    final Iterable<X> origin, final Iterable<Func<X, Boolean>> funcs
) {
```

or like this: 

```java
public Grouped(
    final Iterable<X> origin,
    final Iterable<Func<X, Boolean>> funcs
) {
```

**Comment by amihaiemil** [comment]
@svendiedrichsen Can you also add a unit test for empty iterators (what happens if the given iterators are empty) ?

**Comment by svendiedrichsen** [comment]
@amihaiemil I have made the requested changes and added tests for empty data and functions iterables.

**Comment by amihaiemil** [comment]
@svendiedrichsen You made this PR without being assigned to the Issue. Now, the actual assignee has made another PR with a solution that is a bit different. I will have a look over that one... I've also asked the ticket's author to check these 2 PRs and tell which one is best, but the ticket's assignee has priority, I guess.

**Comment by svendiedrichsen** [comment]
Sorry.

**Comment by 0crat** [comment]
Job [#642](https://github.com/yegor256/cactoos/issues/642) is already in scope

**Comment by 0crat** [comment]
@amihaiemil[/z](https://www.0crat.com/u/amihaiemil) resigned from [#642](https://github.com/yegor256/cactoos/issues/642), since the job is not in scope anymore

**Comment by 0crat** [comment]
The job [#642](https://github.com/yegor256/cactoos/issues/642) is now out of scope

**Comment by amihaiemil** [comment]
@svendiedrichsen 0crat should actually warn you about this situation, when you open a PR for an issue that you are not assigned to. I will open a ticket about that. Anyway, the ticket's author said the other PR is more suitable :D 

And if you want to work with us, you should apply for 0crat, so you get assigned to tasks formally :)

**Comment by svendiedrichsen** [comment]
@amihaiemil Actually - I don't want to work with you anymore. You care more about your tickets than your software. I have seen this too often. Enjoy your tickets.

**Comment by amihaiemil** [comment]
@svendiedrichsen In general, "caring about tickets" means discipline which is crucial for good software. And in this particular case, it is simply a matter of priority and correctness -- the assignee made a PR (and actually, as the ticket's author said, a better one). 

Should the PM have punished that guy? There is currently no mechanism of passing tasks around, if someone rejects or doesn't finish a task, their reputation is lowered. See the ticket I openet about this,  it's linked above...

**Comment by amihaiemil** [comment]
@svendiedrichsen This being said, I guess you are right. We care a lot about tickets -- about just as much some other people care about senseless meetings and phone calls :)
